src/drivers/video/fb.c

```diff
@@ -71,34 +71,33 @@ static void fb_ops_fixup(struct fb_ops *ops) {
 
 struct fb_info *fb_create(const struct fb_ops *ops, char *map_base, size_t map_size) {
 	struct fb_dev *dev;
-	struct fb_info *info;
+	struct fb_info *info = NULL;
+
+	assert(ops);
 
 	mutex_lock(&fb_static);
 	{
 		dev = pool_alloc(&fb_pool);
-		if (dev) {
-			info = &dev->info;
-			info->id = fb_count++;
-			dlist_init(&dev->link);
-			dlist_add_next(&dev->link, &fb_list);
-		} else{
-			info = NULL;
+		if (!dev) {
+			log_error("Failed to create framebuffer");
+			goto out;
 		}
-	}
-	mutex_unlock(&fb_static);
 
-	if (info) {
+		info = &dev->info;
+
+		info->id = fb_count++;
+		dlist_init(&dev->link);
+		dlist_add_next(&dev->link, &fb_list);
 		info->screen_base = map_base;
 		info->screen_size = map_size;
 
 		memcpy(&info->ops, ops, sizeof(struct fb_ops));
 		fb_ops_fixup(&info->ops);
-
-
-		/* FIXME probably should be in open/start */
 		fb_update_current_var(info);
+		fb_devfs_create(info, map_base, map_size);
 	}
-	fb_devfs_create(ops, map_base, map_size);
+out:
+	mutex_unlock(&fb_static);
```

Comment:
mutex_lock/mutex_unlock may be  used safely during the memory allocation because further used only the pointer to  the allocated memory 
`info = &dev->info;`

Reply:
Problem may occur if `fb_lookup()` is called before intialization is complete (for example, if `dlist_add_next` is not finished yet or `info->ops` are not initialized yet).

Same problems for calling `fb_create()` in another thread at the same time.